### PR TITLE
Issue #SB-1720 feat:Announcement Outbox

### PR DESCRIPTION
**Pull request by souravdey091**

### CODE REVIEW

File: src/app/migration/src/src/app/services/pagination/pagination.service.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import * as _ from 'lodash';
+
+export class PagerService {
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/services/pagination/pagination.service.ts]: Change this to pagination, keep consistent

File: src/app/migration/src/src/app/services/pagination/pagination.service.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+   getPager(totalItems: number, currentPage: number = 1, pageSize: number = 10) {
+       // calculate total pages
+       const totalPages = Math.ceil(totalItems / pageSize);
+
+       let startPage: number, endPage: number;
+       if (totalPages <= 5) {
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/services/pagination/pagination.service.ts]: Refactor and simplify and page strip size should be configurable

File: src/app/migration/src/src/app/services/announcement/announcement.service.ts [hunk 3]
```diff
@@ -2,12 +2,128 @@ import { DataService } from './../data/data.service';
...
+  }
+
+  /**
+  * Method to make delete api call
+  *
+  * @param {RequestParam} requestParam Request object needed for delete API call
+  */
+  deleteAnnouncement(requestParam: RequestParam) {
+    const option = {
+      url: urlConFig.URLS.ANNOUNCEMENT.CANCEL,
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/services/announcement/announcement.service.ts]: urlConFig to urlConfig

File: src/app/migration/src/src/app/services/announcement/announcement.service.ts [hunk 4]
```diff
@@ -2,12 +2,128 @@ import { DataService } from './../data/data.service';
...
+  *
+  * @param {RequestParam} requestParam Request object needed for read API call
+  */
+  readAnnouncement(requestParam: RequestParam) {
+    const option = {
+      url: urlConFig.URLS.ANNOUNCEMENT.READ,
+      data: {
+        'request': {
+          'announcementId': requestParam.announcementId,
+          'channel': 'web'
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/services/announcement/announcement.service.ts]: move this to config

File: src/app/migration/src/src/app/services/announcement/announcement.service.ts [hunk 5]
```diff
@@ -2,12 +2,128 @@ import { DataService } from './../data/data.service';
...
+          'limit': requestParam.limit,
+          'offset': (requestParam.pageNumber - 1) * requestParam.limit
+        }
+      }
+    };
+    return this.post(option);
+  }
+
+  /**
+  * Method to make api call to get outbox data
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/services/announcement/announcement.service.ts:54]: Documentation need to be updated

File: src/app/migration/src/src/app/services/announcement/announcement.service.ts [hunk 6]
```diff
@@ -2,12 +2,128 @@ import { DataService } from './../data/data.service';
 import { Injectable } from '@angular/core';
 import { HttpClient } from '@angular/common/http';
 import * as  urlConfig from './../../config/url.config.json';
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/services/announcement/announcement.service.ts]: Move this to service in sharable module

File: src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+  /**
+	 * To show / hide delete confirm box
+	 */
+  showDeleteModal = false;
+
+  /**
+	 * Contains page limit of outbox list
+	 */
+  pageLimit = 25;
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.ts]: Move this to configuration and similar should be moved to configuration

File: src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        spyOn(announcementService, 'getOutboxData').and.callFake(() => Observable.throw({}));
+        component.renderOutbox(10, 1);
+        expect(component.showLoader).toBe(false);
+        expect(component.showError).toBe(true);
+    }));
+
+    it('should call setpage method', () => {
+        component.pager = {};
+        component.pager.totalPages = 0;
+        component.setPage(1);
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.spec.ts]: without assestions no test should be there

File: src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.html [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      </div>
+    </div>
+  </div>
+  <div class="ui grid" *ngIf="showLoader">
+    <div class="one wide column"></div>
+    <div class="ten wide column">
+      <app-loader></app-loader>
+    </div>
+  </div>
+  <div class="ui grid" *ngIf="!showDataDiv && !showLoader">
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.html]: remove showDataDiv and replace it with outBoxList.length === 0

File: src/app/migration/src/src/app/modules/announcement/announcement.module.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { AnnouncementRoutingModule } from './announcement-routing.module';
+import { SuiModule } from 'ng2-semantic-ui';
+import { OutboxComponent } from './components/outbox/outbox.component';
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/modules/announcement/announcement.module.ts]: import from index.ts

**Comment by HarishGangula** [comment]
1.  All files naming convention should be correct
2. test cases should contain assertions
3. deletePopup component should be added with proper route
4. documentation, html files should be refactored

**Review by HarishGangula — CHANGES_REQUESTED** [review]

File: src/app/migration/src/src/app/modules/announcement/announcement.module.ts (hunk 10 above)
Comment (souravdey091) [reply src/app/migration/src/src/app/modules/announcement/announcement.module.ts]: done

File: src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.html (hunk 9 above)
Comment (souravdey091) [reply src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.html]: done

File: src/app/migration/src/src/app/services/pagination/pagination.service.ts (hunk 1 above)
Comment (souravdey091) [reply src/app/migration/src/src/app/services/pagination/pagination.service.ts]: done

File: src/app/migration/src/src/app/modules/announcement/components/delete/delete.component.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+      }
+    );
+  }
+
+  /**
+   * This method helps to redirect to the previous
+   * page, i.e, outbox listing page with proper page number
+	 *
+	 */
+  redirect() {
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/modules/announcement/components/delete/delete.component.ts:103]: Move this to view

File: src/app/migration/src/src/app/modules/announcement/components/delete/delete.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    const params = { data: { 'request': { 'announcementId': 'fa355310-0b09-11e8-93d1-2970a259a0ba' } } };
+    announcementService.deleteAnnouncement(params).subscribe(
+      apiResponse => {
+        expect(apiResponse.responseCode).toBe('OK');
+        expect(apiResponse.result.status).toBe('cancelled');
+        expect(apiResponse.params.status).toBe('successful');
+      }
+    );
+    fixture.detectChanges();
+    expect(component.showError).toBe(false);
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/modules/announcement/components/delete/delete.component.spec.ts]: Add proper assertions

File: src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        const params = { pageNumber: 2, limit: 1 };
+        announcementService.getOutboxData(params).subscribe(
+            apiResponse => {
+                expect(apiResponse.responseCode).toBe('OK');
+                expect(apiResponse.result.count).toBe(1169);
+                expect(apiResponse.params.status).toBe('successful');
+            }
+        );
+        fixture.detectChanges();
+        expect(component.showLoader).toBe(false);
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/modules/announcement/components/outbox/outbox.component.spec.ts]: 1. Change variable name
2. assert UI also

File: src/app/migration/src/src/app/services/announcement/announcement.service.ts [hunk 14]
```diff
@@ -1,13 +1,137 @@
...
+import 'rxjs/add/operator/map';
+import 'rxjs/add/observable/throw';
+import { Observable } from 'rxjs/Observable';
+const urlConfig = (<any>urlConf);
+const pageConfig = (<any>appConfig);
 
+/**
+ * Interface to hold api request pageNumber, limit, announcementId and data
+ */
+interface RequestParam {
```
Comment (HarishGangula) [inline src/app/migration/src/src/app/services/announcement/announcement.service.ts]: Move this interface to file

**Review by HarishGangula — CHANGES_REQUESTED** [review]

**Pull request closed without merging**